site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -2365,6 +2365,78 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
         }
     }
 
+    /**
+     * Route for saving the marks the submitter received for a component
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/change_grade_version", methods: ["POST"])]
+    public function ajaxUpdateGradedVersionForStudent(string $gradeable_id): ?Jsonresponse {
+        $anon_id = $_POST['anon_id'] ?? null;
+        $graded_version = intval($_POST['graded_version'] ?? null);
+        $component_ids = $_POST['component_ids'] ?? [];
+
+        if ($anon_id === null) {
+            return JsonResponse::getFailResponse('Missing anon_id parameter');
+        }
+        if ($graded_version < 1) {
+            return JsonResponse::getFailResponse('Invalid graded_version parameter');
+        }
+
+        if (count($component_ids) < 1) {
+            return JsonResponse::getFailResponse('Missing component_ids parameter');
+        }
+
+        // Get the gradeable
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            return null;
```

Comment:
Should this return a fail response instead of null? Same with other null return values?

Reply:
Technically the tryGetGradables already prints the fail response so by returning a fail response I think we would be sending two fail responses. Or would returning a fail response overwrite that?

Reply:
I believe that most of them have a 'render_json' bool variable, which you can pass to not render JSON from inside the function, and instead return a JsonResponse in this file. 

When looking through the codebase, it may be eventually good to refactor some of these functions to return a JsonResponse (Success or Fail), and then return that value if it is a fail response, as they have more descriptive error messages.

Reply:
Fixed b8fdb7e28ba93ae8e06cc0a8de73cae573df293d